integration/run_test.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/awslabs/soci-snapshotter/util/testutil"
+)
+
+// TestRunMultipleContainers multiple containers at the same time and performs a test in each
+func TestRunMultipleContainers(t *testing.T) {
+
+	tests := []struct {
+		name       string
+		containers []containerImageAndTestFunc
+	}{
+		{
+			name: "Run multiple containers from the same image",
+			containers: []containerImageAndTestFunc{
+				{
+					containerImage: nginxImage,
+					testFunc:       testWebServiceContainer,
+				},
+				{
+					containerImage: nginxImage,
+					testFunc:       testWebServiceContainer,
+				},
+			},
+		}, {
+			name: "Run multiple containers from different images",
+			containers: []containerImageAndTestFunc{
+				{
+					containerImage: nginxImage,
+					testFunc:       testWebServiceContainer,
+				},
+				{
+					containerImage: drupalImage,
+					testFunc:       testWebServiceContainer,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			regConfig := newRegistryConfig()
+			sh, done := newShellWithRegistry(t, regConfig)
+			defer done()
+
+			// Initialize config files for containerd and snapshotter
+			additionalConfig := ""
+			if !isTestingBuiltinSnapshotter() {
+				additionalConfig = proxySnapshotterConfig
+			}
+			containerdConfigYaml := testutil.ApplyTextTemplate(t, `
+version = 2
+
+[plugins."io.containerd.snapshotter.v1.soci"]
+root_path = "/var/lib/soci-snapshotter-grpc/"
+
+[plugins."io.containerd.snapshotter.v1.soci".blob]
+check_always = true
+
+{{.AdditionalConfig}}
+`, struct {
+				AdditionalConfig string
+			}{
+				AdditionalConfig: additionalConfig,
+			})
+
+			// Setup environment
+			if err := testutil.WriteFileContents(sh, defaultContainerdConfigPath, []byte(containerdConfigYaml), 0600); err != nil {
+				t.Fatalf("failed to write %v: %v", defaultContainerdConfigPath, err)
+			}
+			sh.
+				X("update-ca-certificates").
+				Retry(100, "nerdctl", "login", "-u", regConfig.user, "-p", regConfig.pass, regConfig.host)
+
+			rebootContainerd(t, sh, "", "")
+			for _, container := range deduplicateByContainerImage(tt.containers) {
+				// Mirror image
+				copyImage(sh, dockerhub(container.containerImage), regConfig.mirror(container.containerImage))
+				// Pull image, create SOCI index
+				indexDigest := optimizeImage(sh, regConfig.mirror(container.containerImage))
+
+				sh.X("soci", "image", "rpull", "--user", regConfig.creds(), "--soci-index-digest", indexDigest, regConfig.mirror(container.containerImage).ref)
+			}
+
+			// Run the containers
+			for index, container := range tt.containers {
+				image := regConfig.mirror(container.containerImage).ref
+				containerName := "test_" + fmt.Sprint(index) + "_" + makeImageNameValid(container.containerImage)
+				sh.X("soci", "run", "-d", "--rm", "--snapshotter=soci", image, containerName)
+			}
+
+			time.Sleep(2 * time.Second) // FIXME find better solution for "connection refused" from `curl` below too soon
```

Comment:
It looks like this might not be quite long enough, the most recent run failed with:
```
curl: (7) Failed to connect to 127.0.0.1 port 80: Connection refused
```
https://github.com/awslabs/soci-snapshotter/actions/runs/3760427063/jobs/6391131866#step:4:14365


To get this merged, I think we can just increase this timeout for now. As a follow up, it might be worth switching from testing that we make a single request and it succeeds to retrying the request and making sure it succeeds at least once within a timeout.

Reply:
Switched to `curl --retry --retry-connrefused` with what felt like sensible settings